kratos/utilities/assign_unique_model_part_collection_tag_utility.cpp

```diff
@@ -167,6 +171,66 @@ void AssignUniqueModelPartCollectionTagUtility::ComputeTags(
 /***********************************************************************************/
 /***********************************************************************************/
 
+void AssignUniqueModelPartCollectionTagUtility::SetParallelModelPartAndSubModelPartCollectionsAndCombinations(
+                                            IndexStringMapType& rCollections,
+                                            IndexSetIndexMapType& rCombinations,
+                                            IndexType& rTag)
+{
+    auto& r_data_communicator = mrModelPart.GetCommunicator().GetDataCommunicator();
+    const IndexType rank = r_data_communicator.Rank();
+    const IndexType size = r_data_communicator.Size();
+
+    std::vector<std::vector<IndexType>> local_combination_vector_keys;
+    for(auto& key_combination_set : rCombinations) {
+        std::vector<IndexType> key_vector;
+        for (auto it=key_combination_set.first.begin(); it != key_combination_set.first.end(); ++it)
+            key_vector.push_back(*it);
+        local_combination_vector_keys.push_back(key_vector);
+    }
+
+    std::vector<std::vector<IndexType>> global_combination_vector_keys;
+    if (rank>0) {
+        r_data_communicator.Send(local_combination_vector_keys, 0);
+    }
+    else {
+        for (IndexType i_rank = 1; i_rank<size; i_rank++) {
+            std::vector<std::vector<IndexType>> recv_irank_combinations;
+            r_data_communicator.Recv(recv_irank_combinations, i_rank);
+
+            for (auto& key_combination_vector : recv_irank_combinations) {
+                // Convert vector to set
+                const std::set<IndexType> submodel_part_set(key_combination_vector.begin(), key_combination_vector.end());
+                // Check if key exists in the combinations maps, if not, add it
+                if (rCombinations.find(submodel_part_set) == rCombinations.end()) {
+                    rCombinations[submodel_part_set] = 0; //creating new key, value does not matter
+                }
+            }
+        }
+        // Converting back to vector to be broadcasted
+        for(auto& r_key_combination_set : rCombinations) {
+            std::vector<IndexType> key_vector;
+            for (auto it=r_key_combination_set.first.begin(); it != r_key_combination_set.first.end(); ++it)
+                key_vector.push_back(*it);
+            global_combination_vector_keys.push_back(key_vector);
+        }
+    }
```

Comment:
note that this will become slower and slower with more and more ranks, since rank 0 does all the work

probably some proper reduction algorithm would improve this, but is ofc more work
Just pointing it out

Reply:
I see, thanks for the suggestion. I will try to think about it as a future improvement.